pkg/tfgen/convert_cli.go

```diff
@@ -0,0 +1,453 @@
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	hcl2yaml "github.com/pulumi/pulumi-yaml/pkg/pulumiyaml/codegen"
+	hcl2dotnet "github.com/pulumi/pulumi/pkg/v3/codegen/dotnet"
+	hcl2go "github.com/pulumi/pulumi/pkg/v3/codegen/go"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
+	hcl2nodejs "github.com/pulumi/pulumi/pkg/v3/codegen/nodejs"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/pcl"
+	hcl2python "github.com/pulumi/pulumi/pkg/v3/codegen/python"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+var (
+	cliConverterEnabled bool = cmdutil.IsTruthy(os.Getenv("PULUMI_CONVERT"))
+)
+
+// Integrates with `pulumi convert` command for converting TF examples.
+//
+// Pulumi CLI now supprts a handy `pulumi convert` command. This file implements integrating with
+// this command for the purposes of initial conversion of Terraform examples into PCL language. This
+// integration is preferable to linking the functionality in as it allows bridged providers to not
+// build-depend on the TF converter.
+//
+// Note that once examples are converted to PCL, they continue to be processed with in-process
+// target language specific generators to produce TypeScript, YAML, Python etc target code.
+type cliConverter struct {
+	packageName  string                // name of the provider such as "gcp"
+	info         tfbridge.ProviderInfo // provider declaration
+	pluginHost   plugin.Host           // the plugin host for PCL conversion
+	packageCache *pcl.PackageCache     // the package cache for PCL conversion
+
+	hcls map[string]struct{} // set of observed HCL snippets
+
+	generator interface {
+		convertHCL(hcl, path, exampleTitle string, languages []string) (string, error)
+		convertExamplesInner(
+			docs string,
+			path examplePath,
+			stripSubsectionsWithErrors bool,
+			convertHCL func(hcl, path, exampleTitle string,
+				languages []string) (string, error),
+		) string
+	}
+
+	convertExamplesList []struct {
+		docs                       string
+		path                       examplePath
+		stripSubsectionsWithErrors bool
+	}
+
+	currentPackageSpec *pschema.PackageSpec
+
+	pcls map[string]translatedExample // translations indexed by HCL
+	opts []pcl.BindOption             // options cache; do not set
+}
+
+// Represents a partially converted example. PCL is the Pulumi dialect of HCL.
+type translatedExample struct {
+	PCL         string          `json:"pcl"`
+	Diagnostics hcl.Diagnostics `json:"diagnostics"`
+}
+
+// Get or create the cliConverter associated with the Generator.
+func (g *Generator) cliConverter() *cliConverter {
+	if g.cliConverterState != nil {
+		return g.cliConverterState
+	}
+	packageName := string(g.pkg)
+	infoCopy := g.info
+	infoCopy.Name = packageName // correct "google-beta" to "gcp"
```

Comment:
`Provider.Name` is the TF provider name. It should be `"google-beta"`.

Reply:
Well I seem to need that; without it the examples are all incorrect as they render as if the provider is unknown. I'll check again. I think this may have more to do with how the mappings files are named, apparently when passing `--mappings` to `pulumi convert`, the filename is significant and has to identify the provider/package. I might be conflating the two variables here by using info.Name for the filename. Checking up on this.

Reply:
OK I factored out how to compute --mappings names and it's fixed up now.